crates/runtime/src/memory.rs

```diff
@@ -374,7 +431,104 @@ impl RuntimeLinearMemory for ExternalMemory {
         }
     }
 
-    #[cfg(feature = "pooling-allocator")]
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+}
+
+/// For shared memory (and only for shared memory), this lock-version restricts
+/// access when growing the memory or checking its size. This is to conform with
+/// the [thread proposal]: "When `IsSharedArrayBuffer(...)` is true, the return
+/// value should be the result of an atomic read-modify-write of the new size to
+/// the internal `length` slot."
+///
+/// [thread proposal]:
+///     https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md#webassemblymemoryprototypegrow
+#[derive(Clone)]
+pub struct SharedMemory(Arc<SharedMemoryInner>);
+impl SharedMemory {
+    /// Construct a new [`SharedMemory`].
+    pub fn new(ty: wasmtime_types::Memory, tunables: &Tunables) -> Result<Self> {
+        assert!(ty.shared);
+        let plan = MemoryPlan::for_memory(ty, tunables);
+        let (minimum_bytes, maximum_bytes) = Memory::limit_new(&plan, None)?;
+        let mut mmap_memory = MmapMemory::new(&plan, minimum_bytes, maximum_bytes, None)?;
+        let def = LongTermVMMemoryDefinition(mmap_memory.vmmemory());
+        let memory: RwLock<Box<dyn RuntimeLinearMemory>> = RwLock::new(Box::new(mmap_memory));
+        Ok(Self(Arc::new(SharedMemoryInner { memory, ty, def })))
+    }
+
+    /// Wrap an existing [Memory] with the locking provided by a [SharedMemory].
+    pub fn wrap(mut memory: Box<dyn RuntimeLinearMemory>, ty: wasmtime_types::Memory) -> Self {
+        assert!(
+            memory.as_any_mut().type_id() != std::any::TypeId::of::<SharedMemory>(),
+            "cannot re-wrap a shared memory"
+        );
+        let def = LongTermVMMemoryDefinition(memory.vmmemory());
+        Self(Arc::new(SharedMemoryInner {
+            memory: RwLock::new(memory),
+            ty,
+            def,
+        }))
+    }
+
+    /// Return the memory type for this [`SharedMemory`].
+    pub fn ty(&self) -> wasmtime_types::Memory {
+        self.0.ty
+    }
+
+    /// Convert this shared memory into a [`Memory`].
+    pub fn as_memory(self) -> Memory {
+        Memory(Box::new(self))
+    }
+
+    /// Return a mutable pointer to the shared memory's [VMMemoryDefinition].
+    pub fn vmmemory_ptr_mut(&mut self) -> *mut VMMemoryDefinition {
+        &self.0.def.0 as *const _ as *mut _
+    }
+
+    /// Return a pointer to the shared memory's [VMMemoryDefinition].
+    pub fn vmmemory_ptr(&self) -> *const VMMemoryDefinition {
+        &self.0.def.0 as *const _
+    }
+}
+
+struct SharedMemoryInner {
+    memory: RwLock<Box<dyn RuntimeLinearMemory>>,
```

Comment:
Since the outer `SharedMemory` is boxed up as well, could this perhaps be `SharedMemoryInner<T>` and the memory here is `memory: RwLock<T>`?

Reply:
Or actually I see now that the downcast into a shared memory might be difficult with that, so could this otherwise be hardcoded as an `MmapMemory`?